File: actionbar/src/main/java/mobi/cangol/mobile/actionbar/view/ClearableEditText.java
package mobi.cangol.mobile.actionbar.view;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.v7.widget.AppCompatEditText;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.MotionEvent;
import android.view.View;

import mobi.cangol.mobile.actionbar.R;


/**
 * @author Cangol
 */
public class ClearableEditText extends AppCompatEditText {

    Drawable imgX;

    public ClearableEditText(Context context) {
        super(context);
        initViews();
    }

    public ClearableEditText(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        initViews();
    }

    public ClearableEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
        initViews();
    }

    void initViews() {
        TypedValue typedValue = new TypedValue();
        getContext().getTheme().resolveAttribute(R.attr.actionbar_clear, typedValue, true);
        imgX = getResources().getDrawable(typedValue.resourceId);

        // Set bounds of our X button
        imgX.setBounds(0, 0, imgX.getIntrinsicWidth(), imgX.getIntrinsicHeight());

        // There may be initial text in the field, so we may need to display the button
        manageClearButton();

        this.setOnTouchListener(new OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {

                ClearableEditText et = ClearableEditText.this;

                // Is there an X showing?
                if (et.getCompoundDrawables()[2] == null) return false;
                // Only do this for up touches
                if (event.getAction() != MotionEvent.ACTION_UP) return false;
                // Is touch on our clear button?
                if (event.getX() > et.getWidth() - et.getPaddingRight() - imgX.getIntrinsicWidth()) {
                    et.setText("");
                    removeClearButton();
                }
                return false;
            }
        });

        this.addTextChangedListener(new TextWatcher() {
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable arg0) {
                manageClearButton();
            }

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }
        });
        this.setOnFocusChangeListener(new OnFocusChangeListener() {

            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (!hasFocus) {
                    removeClearButton();
                } else {
                    manageClearButton();
                }
            }
        });
    }

    void manageClearButton() {
        if (this.getText().toString().equals("")) {
            removeClearButton();
        } else if (this.hasFocus()) {
            addClearButton();
        }

    }

    void addClearButton() {
        this.setCompoundDrawables(this.getCompoundDrawables()[0],
                this.getCompoundDrawables()[1],
                getError() == null ? imgX : this.getCompoundDrawables()[2],
                this.getCompoundDrawables()[3]);
    }

    void removeClearButton() {
        this.setCompoundDrawables(this.getCompoundDrawables()[0],
                this.getCompoundDrawables()[1],
                getError() == null ? null : this.getCompoundDrawables()[2],
                this.getCompoundDrawables()[3]);
    }

}

File: actionbar/src/main/java/mobi/cangol/mobile/actionbar/view/ClearableEditText.kt
package mobi.cangol.mobile.actionbar.view

import android.content.Context
import android.graphics.drawable.Drawable
import android.support.v7.widget.AppCompatEditText
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.AttributeSet
import android.util.TypedValue
import android.view.MotionEvent
import android.view.View.OnFocusChangeListener
import android.view.View.OnTouchListener
import mobi.cangol.mobile.actionbar.R


/**
 * @author Cangol
 */
class ClearableEditText : AppCompatEditText {

    private var imgX: Drawable? = null
    private var textWatcher: TextWatcher? = null

    constructor(context: Context) : super(context) {
        initViews()
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        initViews()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        initViews()
    }

    private fun initViews() {
        val typedValue = TypedValue()
        context.theme.resolveAttribute(R.attr.actionbar_clear, typedValue, true)
        imgX = resources.getDrawable(typedValue.resourceId)

        // Set bounds of our X button
        imgX?.setBounds(0, 0, imgX?.intrinsicWidth?:0, imgX?.intrinsicHeight?:0)

        // There may be initial text in the field, so we may need to display the button
        manageClearButton()

        this.setOnTouchListener(OnTouchListener { v, event ->
            val et = this@ClearableEditText

            // Is there an X showing?
            if (et.compoundDrawables[2] == null) return@OnTouchListener false
            // Only do this for up touches
            if (event.action != MotionEvent.ACTION_UP) return@OnTouchListener false
            // Is touch on our clear button?
            if (event.x > et.width - et.paddingRight - (imgX?.intrinsicWidth?:0)) {
                et.setText("")
                removeClearButton()
            }
            false
        })
        textWatcher = object : TextWatcher {
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                //do nothings
            }

            override fun afterTextChanged(arg0: Editable) {
                manageClearButton()
            }

            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                //do nothings
            }
        }
        this.addTextChangedListener(textWatcher)
        this.onFocusChangeListener = OnFocusChangeListener { v, hasFocus ->
            if (!hasFocus) {
                removeClearButton()
            } else {
                manageClearButton()
            }
        }
    }

    private fun manageClearButton() {
        if (TextUtils.isEmpty(text)) {
            removeClearButton()
        } else if (this.hasFocus()) {
            addClearButton()
        }

    }

    private fun addClearButton() {
        this.setCompoundDrawables(this.compoundDrawables[0],
                this.compoundDrawables[1],
                if (error == null) imgX else this.compoundDrawables[2],
                this.compoundDrawables[3])
    }

    private fun removeClearButton() {
        this.setCompoundDrawables(this.compoundDrawables[0],
                this.compoundDrawables[1],
                if (error == null) null else this.compoundDrawables[2],
                this.compoundDrawables[3])
    }

    override fun onDetachedFromWindow() {
        this.removeTextChangedListener(textWatcher)
        super.onDetachedFromWindow()
    }
}